### PR TITLE
issue: 1284069 Add ability to operate w/o SRIOV

### PR DESCRIPTION
**Pull request by liranoz12**

### CODE REVIEW

File: src/vma/dev/ring_bond.h [hunk 1]
```diff
@@ -101,51 +109,72 @@ class ring_bond_eth : public ring_bond
...
+	inline bool is_vf_mode() { return (bool)m_vf_ring; }
+
 protected:
-	virtual void    slave_create(int if_index);
-	virtual void    slave_destroy(int if_index);
+	virtual void slave_create(int if_index);
+
+private:
+	ring_slave*      m_vf_ring;
+	ring_slave*      m_tap_ring;
```
Comment (liranoz12) [inline src/vma/dev/ring_bond.h]: why do we need to save the pointer here and not in the slaves array ? why not ring_simple* and ring_tap* ?

File: src/vma/dev/rfs_uc_tcp_gro.cpp [hunk 2]
```diff
@@ -43,12 +43,19 @@
 #define TCP_H_LEN_TIMESTAMP 8
 
 
-rfs_uc_tcp_gro::rfs_uc_tcp_gro(flow_tuple *flow_spec_5t, ring_simple *p_ring, rfs_rule_filter* rule_filter, uint32_t flow_tag_id) :
+rfs_uc_tcp_gro::rfs_uc_tcp_gro(flow_tuple *flow_spec_5t, ring_slave *p_ring, rfs_rule_filter* rule_filter, uint32_t flow_tag_id) :
 	rfs_uc(flow_spec_5t, p_ring, rule_filter, flow_tag_id),
-	m_p_gro_mgr(&(p_ring->m_gro_mgr)), m_b_active(false), m_b_reserved(false)
+	m_b_active(false), m_b_reserved(false)
 {
+	ring_simple* p_check_ring = dynamic_cast<ring_simple*>(p_ring);
```
Comment (liranoz12) [inline src/vma/dev/rfs_uc_tcp_gro.cpp]: why fatal for ring tap? why not just reciveing ring_simple *p_ring ?

File: src/vma/dev/rfs.cpp [hunk 3]
```diff
@@ -128,11 +128,15 @@ rfs::~rfs()
 		int counter = 0;
 		prepare_filter_detach(counter, true);
 		if (counter == 0) {
-			destroy_ibv_flow();
+			if (m_p_ring->get_type() == RING_SIMPLE) {
```
Comment (liranoz12) [inline src/vma/dev/rfs.cpp]: m_p_ring->get_type() == RING_SIMPLE should move to destroy_ibv_flow( & create_ibv_flow

File: src/vma/dev/ring_tap.cpp [hunk 4]
```diff
@@ -0,0 +1,1213 @@
...
+}
+
+void ring_tap::send_lwip_buffer(ring_user_id_t id, vma_ibv_send_wr* p_send_wqe, bool b_block)
+{
+	NOT_IN_USE(id);
+	auto_unlocker lock(m_lock_ring_tx);
+	mem_buf_desc_t* p_mem_buf_desc = (mem_buf_desc_t*)(p_send_wqe->wr_id);
+	p_mem_buf_desc->lwip_pbuf.pbuf.ref++;
+	vma_wr_tx_packet_attr attr = (vma_wr_tx_packet_attr)((b_block*VMA_TX_PACKET_BLOCK)|VMA_TX_PACKET_L3_CSUM|VMA_TX_PACKET_L4_CSUM);
+	int ret = send_buffer(p_send_wqe, attr);
```
Comment (liranoz12) [inline src/vma/dev/ring_tap.cpp]: redundant

File: src/vma/dev/ring_tap.cpp [hunk 5]
```diff
@@ -0,0 +1,1213 @@
...
+	NOT_IN_USE(id);
+	auto_unlocker lock(m_lock_ring_tx);
+	int ret = send_buffer(p_send_wqe, attr);
+	send_status_handler(ret, p_send_wqe);
+}
+
+void ring_tap::send_lwip_buffer(ring_user_id_t id, vma_ibv_send_wr* p_send_wqe, bool b_block)
+{
+	NOT_IN_USE(id);
+	auto_unlocker lock(m_lock_ring_tx);
```
Comment (liranoz12) [inline src/vma/dev/ring_tap.cpp]: we dont need to lock here

File: src/vma/dev/ring_tap.cpp [hunk 6]
```diff
@@ -0,0 +1,1213 @@
...
+
+		return false;
+	}
+	return p_rfs->rx_dispatch_packet(p_rx_wc_buf_desc, pv_fd_ready_array);
+}
+
+void ring_tap::send_ring_buffer(ring_user_id_t id, vma_ibv_send_wr* p_send_wqe, vma_wr_tx_packet_attr attr)
+{
+	NOT_IN_USE(id);
+	auto_unlocker lock(m_lock_ring_tx);
```
Comment (liranoz12) [inline src/vma/dev/ring_tap.cpp]: we dont need to lock here

File: src/vma/dev/ring_tap.cpp [hunk 7]
```diff
@@ -0,0 +1,1213 @@
...
+	if (p_rx_wc_buf_desc->dec_ref_count() > 1) {
+		// The sink will be responsible to return the buffer to CQ for reuse
+		return true;
+	}
+	// Reuse this data buffer & mem_buf_desc
+	return false;
+}
+
+bool ring_tap::rx_process_buffer(mem_buf_desc_t* p_rx_wc_buf_desc, void* pv_fd_ready_array)
+{
```
Comment (liranoz12) [inline src/vma/dev/ring_tap.cpp]: duplicated code from ring_simple. no need for flowtag

File: src/vma/dev/ring_tap.cpp [hunk 8]
```diff
@@ -0,0 +1,1213 @@
...
+
+void ring_tap::send_status_handler(int ret, vma_ibv_send_wr* p_send_wqe)
+{
+	if (ret && p_send_wqe) {
+		mem_buf_desc_t* p_mem_buf_desc = (mem_buf_desc_t*)(p_send_wqe->wr_id);
+		mem_buf_tx_release(p_mem_buf_desc, true);
+	}
+}
+
+void ring_tap::flow_udp_del_all()
```
Comment (liranoz12) [inline src/vma/dev/ring_tap.cpp]: duplicated code

File: src/vma/dev/ring_tap.cpp [hunk 9]
```diff
@@ -0,0 +1,1213 @@
...
+				buff->sz_data = ret;
+				if (m_vf_ring) {
+					m_vf_ring->rx_process_buffer(buff, pv_fd_ready_array);
+				} else {
+					rx_process_buffer(buff, pv_fd_ready_array);
+				}
+				m_p_ring_stat->n_rx_byte_count += ret;
+				m_p_ring_stat->n_rx_pkt_count++;
+				m_p_ring_stat->tap.n_rx_buffers--;
+			} else {
```
Comment (liranoz12) [inline src/vma/dev/ring_tap.cpp]: remove stats update, why processing using the vf ring ? we should enable sw_csum in this case.
readv ?

File: src/vma/dev/ring_tap.cpp [hunk 10]
```diff
@@ -0,0 +1,1213 @@
...
+			}
+		}
+		m_p_ring_stat->tap.n_rx_buffers = m_rx_pool.size();
+		return true;
+	}
+	return false;
+}
+
+// calling sockinfo callback with RFS bypass
+static inline bool check_rx_packet(sockinfo *si, mem_buf_desc_t* p_rx_wc_buf_desc, void *fd_ready_array)
```
Comment (liranoz12) [inline src/vma/dev/ring_tap.cpp]: no flow tag - redundent

File: src/vma/sock/fd_collection.cpp [hunk 11]
```diff
@@ -197,23 +196,6 @@ void fd_collection::clear()
 		}
 	}
 
-
-	if (!g_is_forked_child && m_p_cma_event_channel) {
```
Comment (liranoz12) [inline src/vma/sock/fd_collection.cpp]: please explain what is the issue here.

File: src/vma/dev/allocator.cpp [hunk 12]
```diff
@@ -92,7 +92,8 @@ void* vma_allocator::alloc_and_reg_mr(size_t size, ib_ctx_handler *p_ib_ctx_h)
 		}
 		else {
 			__log_info_dbg("Huge pages allocation passed successfully");
-			if (!register_memory(size, p_ib_ctx_h, m_non_contig_access_mr)) {
+			if (g_p_ib_ctx_handler_collection->get_ib_cxt_list() &&
```
Comment (liranoz12) [inline src/vma/dev/allocator.cpp]: i dont agree with the check. register_memory will do nothing if the list is empty
get_ib_cxt_list() should stay as is.

File: src/vma/dev/ring_bond.h [hunk 13]
```diff
@@ -79,11 +86,12 @@ class ring_bond : public ring {
 	int 			socketxtreme_poll(struct vma_completion_t *vma_completions, unsigned int ncompletions, int flags);
 #endif // DEFINED_SOCKETXTREME		
 	virtual void    slave_create(int if_index) = 0;
-	virtual void    slave_destroy(int if_index) = 0;
+	virtual void    slave_destroy(int if_index);
 protected:
 	void			update_rx_channel_fds();
 	void			popup_active_rings();
 	ring_slave_vector_t     m_bond_rings;
+	std::vector<struct flow_sink_t> m_rx_flows;
```
Comment (liranoz12) [inline src/vma/dev/ring_bond.h]: can be done using the rfs maps

File: src/vma/dev/ring_tap.cpp [hunk 14]
```diff
@@ -0,0 +1,1213 @@
...
+int ring_tap::send_buffer(vma_ibv_send_wr* wr, vma_wr_tx_packet_attr attr)
+{
+	int ret = 0;
+//	mem_buf_desc_t* buff = (mem_buf_desc_t*)(p_send_wqe->wr_id);
+	int i = 0;
+
+	NOT_IN_USE(attr);
+
+	for (i = 0; i < wr->num_sge; i++) {
+		ret = orig_os_api.write(m_tap_fd, (const void *)wr->sg_list[i].addr, wr->sg_list[i].length);
```
Comment (liranoz12) [inline src/vma/dev/ring_tap.cpp]: please use writev

File: src/vma/dev/ring_tap.cpp [hunk 15]
```diff
@@ -0,0 +1,1213 @@
...
+		ring_logfunc("Out of mem_buf_desc from RX free pool for internal object pool");
+		return false;
+	}
+
+	m_p_ring_stat->tap.n_rx_buffers = m_rx_pool.size();
+
+	return true;
+}
+
+bool ring_tap::request_more_tx_buffers()
```
Comment (liranoz12) [inline src/vma/dev/ring_tap.cpp]: duplicated code from ring_simple

File: src/vma/dev/ring_tap.cpp [hunk 16]
```diff
@@ -0,0 +1,1213 @@
...
+		next->lwip_pbuf.pbuf.ref = 1;
+		n_num_mem_bufs--;
+	}
+
+	m_lock_ring_tx.unlock();
+
+	return head;
+}
+
+int ring_tap::mem_buf_tx_release(mem_buf_desc_t* buff_list, bool b_accounting, bool trylock)
```
Comment (liranoz12) [inline src/vma/dev/ring_tap.cpp]: duplicated code

File: src/vma/dev/ring_tap.cpp [hunk 17]
```diff
@@ -0,0 +1,1213 @@
...
+			this, m_sysvar_qp_compensation_level, 0);
+	if (!res) {
+		ring_logfunc("Out of mem_buf_desc from TX free pool for internal object pool");
+		return false;
+	}
+
+	return true;
+}
+
+mem_buf_desc_t* ring_tap::mem_buf_tx_get(ring_user_id_t id, bool b_block, int n_num_mem_bufs)
```
Comment (liranoz12) [inline src/vma/dev/ring_tap.cpp]: duplicated code

File: src/vma/sock/sockinfo_udp.cpp [hunk 18]
```diff
@@ -1870,12 +1873,36 @@ ssize_t sockinfo_udp::tx(const tx_call_t call_type, const iovec* p_iov, const ss
 			b_blocking = false;
 
 		if (likely(p_dst_entry->is_valid())) {
+			/* Purpose of this condition to avoid redirect traffic to OS
```
Comment (liranoz12) [inline src/vma/sock/sockinfo_udp.cpp]: we should send the data from the tap also for udp, can be removed

File: src/vma/dev/ib_ctx_handler_collection.cpp [hunk 19]
```diff
@@ -95,8 +95,10 @@ void ib_ctx_handler_collection::update_tbl()
 		throw_vma_exception("No IB capable devices found!");
 	}
 	if (!num_devices) {
-		ibchc_logdbg("No IB capable devices found!");
-		throw_vma_exception("No IB capable devices found!");
+		vlog_printf(VLOG_WARNING, "*************************************************************\n");
```
Comment (liranoz12) [inline src/vma/dev/ib_ctx_handler_collection.cpp]: change to debug

File: src/vma/sock/sockinfo_tcp.cpp [hunk 20]
```diff
@@ -4409,12 +4409,7 @@ void sockinfo_tcp::tcp_tx_pbuf_free(void* p_conn, struct pbuf *p_buff)
 		mem_buf_desc_t * p_desc = (mem_buf_desc_t *)p_buff;
 
 		//potential race, ref is protected here by tcp lock, and in ring by ring_tx lock
-		if (likely(p_desc->lwip_pbuf_get_ref_count()))
```
Comment (liranoz12) [inline src/vma/sock/sockinfo_tcp.cpp]: we should not remove the error message. but i agree that we can optimize this section

File: src/vma/dev/allocator.cpp (hunk 12 above)
Comment (rafiw) [reply src/vma/dev/allocator.cpp]: why not to move this check to register function?

**Pull request merged**